packages/opentelemetry-exporter-datadog/src/transform.ts

```diff
@@ -0,0 +1,299 @@
+    // we can infer a type by using the status code and also the non spec `<library>.error_name` attribute
+    const possibleType = inferErrorType(span);
+    ddSpanBase.setTag('error', 1);
+    ddSpanBase.setTag('error.msg', span.status.message);
+
+    if (possibleType) {
+      ddSpanBase.setTag('error.type', possibleType);
+    }
+  }
+
+  // set span kind
+  if (DD_SPAN_KIND_MAPPING[span.kind]) {
+    ddSpanBase.setTag('span.kind', DD_SPAN_KIND_MAPPING[span.kind]);
+  }
+
+  // set span type
+  // span.instrumentationLibrary.name is not in v0.9.0 but has been merged
+  if (getInstrumentationName(span)) {
+    if (DD_SPAN_TYPE_MAPPING[getInstrumentationName(span) || '']) {
+      ddSpanBase.setTag(
+        'span.type',
+        DD_SPAN_TYPE_MAPPING[getInstrumentationName(span) || '']
+      );
+    }
+  }
+
+  // set env tag
+  if (env) {
+    ddSpanBase.setTag(ENV_KEY, env);
+  }
+
+  // set origin and version on root span only
+  if (span.parentSpanId === undefined) {
+    const origin = createOriginString(span);
+    if (origin) {
+      ddSpanBase.setTag(DD_ORIGIN, origin);
+    }
+    if (version) {
+      ddSpanBase.setTag(VERSION_KEY, version);
+    }
+  }
+
+  // set span attibutes as tags - takes precedence over env vars
+  // span tags should be strings
+  for (const [key, value] of Object.entries(span.attributes)) {
+    if (value !== undefined) {
+      ddSpanBase.setTag(key, value);
+    }
+  }
+
+  // filter for internal requests to trace-agent and set sampling rate
+  const samplingRate = getSamplingRate(span);
+  const internalRequest = isInternalRequest(span);
+
+  if (internalRequest) {
+    ddSpanBase.setTag(SAMPLE_RATE_METRIC_KEY, USER_REJECT);
+  } else if (samplingRate !== undefined) {
+    ddSpanBase.setTag(SAMPLE_RATE_METRIC_KEY, samplingRate);
+  }
+
+  // set span duration and mark as finished span so that it is exported
+  const duration =
+    hrTimeToMilliseconds(span.endTime) - hrTimeToMilliseconds(span.startTime);
+  ddSpanBase._duration = duration;
+  ddSpanBaseContext._isFinished = true;
+
+  return ddSpanBase;
+}
+
+function getTraceContext(span: ReadableSpan): typeof Span[] {
+  return [
+    id(span.spanContext['traceId']),
+    id(span.spanContext['spanId']),
+    span.parentSpanId ? id(span.parentSpanId) : null,
+  ];
+}
+
+function createDefaultTags(tags: string | undefined): object {
+  // Parse a string of tags typically provided via environment variables.
+  // The expected string is of the form: "key1:value1,key2:value2"
+
+  const tagMap = tags
+    ? tags
+        .split(',')
+        .reduce((tags: { [key: string]: string }, kvPair: string) => {
+          const kvTuple = kvPair.split(':');
+          tags[kvTuple[0]] = kvTuple[1];
+          return tags;
+        }, {})
+    : {};
+
+  // ensure default tag env var or  arg is not malformed
+  if (
+    Object.keys(tagMap).indexOf('') >= 0 ||
+    Object.values(tagMap).indexOf('') >= 0 ||
```

Comment:
please cache result of Object.values(tagMap) as it is used twice here, unless you go with previous solution

Reply:
👍 went with a variant of the below (using object.entries I can do it in one partial loop)